sdks/python/apache_beam/io/filebasedsink.py

```diff
@@ -188,39 +190,51 @@ def _get_final_name(self, shard_num, num_shards):
         self.file_name_suffix.get()
     ])
 
+  @check_accessible(['file_path_prefix', 'file_name_suffix'])
+  def _get_final_name_glob(self, num_shards):
+    return ''.join([
+      self.file_path_prefix.get(),
+      self.shard_name_glob_format % dict(num_shards=num_shards),
+      self.file_name_suffix.get()
+    ])
+
   def pre_finalize(self, init_result, writer_results):
-    writer_results = sorted(writer_results)
-    num_shards = len(writer_results)
-    existing_files = []
-    for shard_num in range(len(writer_results)):
-      final_name = self._get_final_name(shard_num, num_shards)
-      if FileSystems.exists(final_name):
-        existing_files.append(final_name)
-    if existing_files:
-      logging.info('Deleting existing files in target path: %d',
-                   len(existing_files))
-      FileSystems.delete(existing_files)
+    num_shards = len(list(writer_results))
+    dst_glob = self._get_final_name_glob(num_shards)
+    dst_glob_files = [file_metadata.path
+                      for mr in FileSystems.match([dst_glob])
+                      for file_metadata in mr.metadata_list]
 
-  @check_accessible(['file_path_prefix'])
-  def finalize_write(self, init_result, writer_results,
-                     unused_pre_finalize_results):
-    writer_results = sorted(writer_results)
-    num_shards = len(writer_results)
+    if dst_glob_files:
+      logging.info('Deleting existing files in target path: %d',
+                   len(dst_glob_files))
+      FileSystems.delete(dst_glob_files)
+
+  def _check_state_for_finalize_write(self, writer_results, num_shards):
```

Comment:
Please add comments describing this and other util methods added.

Reply:
done